pkg/webhook/stats_reporter.go

```diff
@@ -55,23 +55,20 @@ func newStatsReporter() (StatsReporter, error) {
 }
 
 // Captures req count metric, recording the count and the duration
-func (r *reporter) ReportAdmissionRequest(response admissionReqRes, d time.Duration) error {
-	ctx, err := tag.New(
-		r.ctx,
-		tag.Insert(admissionStatusKey, string(response)),
-	)
-	if err != nil {
-		return err
-	}
-
-	return r.report(ctx, responseTimeInSecM.M(d.Seconds()))
+func (r *reporter) ReportAdmissionRequest(response requestResponse, d time.Duration) error {
+	return r.reportRequest(response, d, admissionStatusKey)
 }
 
 // Captures req count metric, recording the count and the duration
-func (r *reporter) ReportMutationRequest(response mutationResponse, d time.Duration) error {
+func (r *reporter) ReportMutationRequest(response requestResponse, d time.Duration) error {
```

Comment:
We should create a separate metric: `mutation_request_count`, the equivalent of `request_count` but for mutation.

Another option would be to add a separate tag, called "webhook" that returns either "mutation" or "validation"

The second is probably cleaner, the first is backwards-compatible.

Reply:
If we are considering changing at the risk of backward compatibility, I would vote for `validating_request_count` to replace `request_count` and `validating_requestDurationMetricName` to replace `requestDurationMetricName`. WDYT?

Reply:
It would probably be less painful for users if we kept `request_count` and all and added special `mutation_` metric names if we go the two separate metrics route.

This is because users may have developed alerts and aggregation rules for prometheus that depend on the old metric names.

If we want to migrate, maybe we should have some sort of deprecation pathway?

Reply:
+1 on deprecation

should we publish all 3 (no prefix, validation and mutation) and then deprecate non prefix (which is same as validation)?

Reply:
That works for me

Reply:
We'll want to update the docstrings on the old metric and add release notes saying its deprecated.

Maybe add a comment giving us a time after which the metric can be removed.

Reply:
+1 on 3 and deprecate `request_count` later

Reply:
Added validation_request_count/ validation_request_duration_seconds
The validating webhook will now report both: [here](https://github.com/open-policy-agent/gatekeeper/pull/976/files#diff-d3c8e52b699d3596d1057b82890a2ed8624445f1ee8f6d3d8a532e13766b99a8R149)
Please confirm this is what we want.

Reply:
Added #1010 to track this

Reply:
I'm late to this discussion but have just been going over the metrics code.
Backwards compatibility aside, I'm not sure separate metrics for separate webhooks is internally consistent. Controller-runtime already exposes a number of metrics for webhooks automatically, and they use labels for distinguishing between them:

```
# HELP controller_runtime_webhook_requests_in_flight Current number of admission requests being served.
# TYPE controller_runtime_webhook_requests_in_flight gauge
controller_runtime_webhook_requests_in_flight{webhook="/v1/admit"} 0
controller_runtime_webhook_requests_in_flight{webhook="/v1/admitlabel"} 0
```

Additionally, with a single metric it is easier to perform aggregation across all webhooks when desired.

Do we have a sense of how we want to consume these metrics and what kind of questions they will help answer? That might help understand the aggregation use case better.

Reply:
Per my original comment, I agree the extra label/tag solution is cleaner. I also don't see a downside internally as I don't think there is much difference between a new metric and a new tag. New tag may actually have slightly less overhead.

My biggest concern would be people who may have built up Prometheus alerts and/or aggregation rules around the metrics as they currently exist. Does adding data from mutation affect how they've tuned their alerts? What are the backwards compatibility expectations for metrics?